### PR TITLE
error handling & formatting

### LINKED ISSUE #139: onError handler
- Being able to log errors

### LINKED ISSUE #144: formatError()-handler

### PR DESCRIPTION
**Pull request by KATT**
- closes #139 
- closes #144 

see 

- https://www-git-feature-err.tmp.trpc.io/docs/error-handling
- https://www-git-feature-err.tmp.trpc.io/docs/error-formatting

---

changes:

- add onError-fn
- add formatError-fn
- now passing procedure type to middlewares
- managed to get rid of `TMiddlewares`-generic
- update error handling
  - all server errors are `TRPCError`-instances
  - client errors get a `.shape`-object that is used is reflected from the `formatError`-fn

### CODE REVIEW

File: www/docs/server/error-formatting.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+```tsx
+
+function MyComponent() {
+  const mutation = hooks.useMutation('addPost');
+
+  useEffect(() => {
+    mutation.mutate({ title: 123 as any });
+  }, []);
+
+  if (mutation.error mutation.error.shape.zodError) {
```
Comment (cyrus-za) [inline www/docs/server/error-formatting.md:62]:
```suggestion
  if (mutation.error?.shape?.zodError) {
```

**Review by cyrus-za — CHANGES_REQUESTED** [review]
this is awesome work! Thanks for putting the time in for getting it out. Check the uncommented code. Other than that it looks amazinf

File: examples/next-ssg-chat/pages/api/trpc/[trpc].ts [hunk 2]
```diff
@@ -127,4 +127,7 @@ export default trpcNext.createNextApiHandler({
   createContext,
   teardown: () => prisma.$disconnect(),
   transformer: superjson,
+  // onError({ error }) {
+  //   console.error('Error:', error);
+  // },
```
[the remark this answers is not in the document]
Comment (KATT) [reply examples/next-ssg-chat/pages/api/trpc/[trpc].ts:132]: it breaks the vercel builds if i uncomment that

[the remark this answers is not in the document]
Comment (KATT) [reply examples/next-ssg-chat/pages/api/trpc/[trpc].ts:132]: as the vercel builds run on the latest published version

File: examples/next-prisma-todomvc/pages/api/trpc/[trpc].ts [hunk 3]
```diff
@@ -20,7 +20,18 @@ export type Context = trpc.inferAsyncReturnType<typeof createContext>;
...
+  // .formatError(({ defaultShape, error }) => {
+  //   return {
+  //     ...defaultShape,
+  //     zodError:
+  //       error.code === 'BAD_USER_INPUT' &&
+  //       error.originalError instanceof ZodError
+  //         ? error.originalError.flatten()
+  //         : null,
+  //   };
+  // })
```
[the remark this answers is not in the document]
Comment (KATT) [reply examples/next-prisma-todomvc/pages/api/trpc/[trpc].ts:33]: see other comment

**Pull request merged**